### PR TITLE
Issue4550 alternative master

### PR DESCRIPTION
**Pull request by soul2zimate**
As I mentioned in #issue4550

javaserverfaces/mojarra@4c74d1f  changed `submittedValue` from `null` to `""` if `javax.faces.INTERPRET_EMPTY_STRING_SUBMITTED_VALUES_AS_NULL` is configured to true
```
            if (myConsiderEmptyStringNull(context)) {
                // JAVASERVERFACES_SPEC_PUBLIC-1329: If the EMPTY_STRING_SUBMITTED_VALUES_AS_NULL
                // config is set, ensure that logic gets a chance to be executed
                // in UIInput.processValidators().
                if (null == submittedValue) {
                    setSubmittedValue("");
                }
            }
```
And it essentially makes a difference in `validate` [method](https://github.com/eclipse-ee4j/mojarra/blob/2.3.14-RELEASE/impl/src/main/java/javax/faces/component/UIInput.java#L982-#L991) in UIInput.java

```
        // Submitted value == null means "the component was not submitted
        // at all".
        Object submittedValue = getSubmittedValue();
        if (submittedValue == null) {
            if (isRequired() && isSetAlwaysValidateRequired(context)) {
                // continue as below
            } else {
                return;
            }
        }
```
If `submittedValue` is null, it returns immediately as it considers `the component was not submitted at all`.

However, `submittedValue` becomes `""`, so it continues to executed in order to validate and set the new values.

Previous fix caused another issue as reported in https://github.com/eclipse-ee4j/mojarra/issues/4704. Here, the new fix just address the JAVASERVERFACES_SPEC_PUBLIC-1329 validation requirement inside UIInput.java rather than UIViewParameter.java. If javax.faces.INTERPRET_EMPTY_STRING_SUBMITTED_VALUES_AS_NULL is enabled, for null value input, it still give a chance to validate "" value in UIInput.validate() method, then return immediately.

3.0 branch PR: https://github.com/eclipse-ee4j/mojarra/pull/4707
2.3 branch PR: https://github.com/eclipse-ee4j/mojarra/pull/4708

WFLY issue: https://issues.redhat.com/browse/WFLY-13497

### CODE REVIEW

**Comment by soul2zimate** [comment]
@arjantijms Could you please review ?

**Comment by arjantijms** [comment]
@BalusC could you take a look at this first? I remember this is quite subtle, and you were most up to date about it.

File: test/javaee6/viewParamNullValueAjax/src/main/java/com/sun/faces/test/javaee6/viewParamNullValueAjax/TestView.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+package com.sun.faces.test.javaee6.viewParamNullValueAjax;
+
+import java.io.Serializable;
+
+import javax.enterprise.inject.Model;
+import javax.faces.view.ViewScoped;
+import javax.inject.Named;
+
+@Model
```
Comment (BalusC) [inline test/javaee6/viewParamNullValueAjax/src/main/java/com/sun/faces/test/javaee6/viewParamNullValueAjax/TestView.java:25]: Remove this. It's a stereotype for `@Named @RequestScoped`. But it's being overridden with `@ViewScoped @Named` in following lines.

File: test/servlet31/faceletsEmptyAsNull/src/test/java/com/sun/faces/test/servlet31/faceletsemptyasnull/Issue2827IT.java [hunk 2]
```diff
@@ -144,10 +144,10 @@ public void testValidateEmptyFields() throws Exception {
...
 
             pageAsText = page.asText();
-            assertTrue(pageAsText.contains("VC1 Fired: true"));
-            assertTrue(pageAsText.contains("VC2 Fired: true"));
-            assertTrue(pageAsText.contains("String model set with null: false"));
-            assertTrue(pageAsText.contains("Integer model set with null: false"));
+            assertTrue(pageAsText.contains("VC1 Fired: false"));
+            assertTrue(pageAsText.contains("VC2 Fired: false"));
+            assertTrue(pageAsText.contains("String model set with null: true"));
+            assertTrue(pageAsText.contains("Integer model set with null: true"));
```
Comment (BalusC) [inline test/servlet31/faceletsEmptyAsNull/src/test/java/com/sun/faces/test/servlet31/faceletsemptyasnull/Issue2827IT.java:150]: Was the original test wrong? This wasn't explained in conversation?

File: test/javaee6/viewParamNullValueAjax/src/main/java/com/sun/faces/test/javaee6/viewParamNullValueAjax/TestView.java (hunk 1 above)
Comment (soul2zimate) [reply test/javaee6/viewParamNullValueAjax/src/main/java/com/sun/faces/test/javaee6/viewParamNullValueAjax/TestView.java]: Thank you for the review, I rebased and removed that annotation.

File: test/servlet31/faceletsEmptyAsNull/src/test/java/com/sun/faces/test/servlet31/faceletsemptyasnull/Issue2827IT.java (hunk 2 above)
Comment (soul2zimate) [reply test/servlet31/faceletsEmptyAsNull/src/test/java/com/sun/faces/test/servlet31/faceletsemptyasnull/Issue2827IT.java:150]: I thought the test was wrong when I prepared https://github.com/eclipse-ee4j/mojarra/pull/4643, but I was wrong. This pull request reverts it back.

**Comment by soul2zimate** [comment]
Hi @arjantijms, since this was approved by @BalusC, Could you review and merge this please ?

**Comment by arjantijms** [comment]
@soul2zimate absolutely, though I do need to ask you to sign off the commit. There's now this error at eclipse:

```
Chao Wang (chaowan@****.com) did not include the "Signed-off-by footer" which is required for all commits made by a contributor.
```

**Comment by soul2zimate** [comment]
sure, I have  signed off all commits now.

**Comment by arjantijms** [comment]
Thx, the button is now green :)

**Pull request merged**